{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Deduper Script Testing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inline Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1', '2', '3'}\n",
      "Yes 2\n"
     ]
    }
   ],
   "source": [
    "x = set()\n",
    "x.add(\"1\")\n",
    "x.add(\"2\")\n",
    "x.add(\"3\")\n",
    "print(x)\n",
    "\n",
    "if \"2\" in x:\n",
    "    print(\"Yes 2\")\n",
    "if \"4\" in x:\n",
    "    print(\"Yes 4\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'a': [1, 2, 3], 'b': [4, 5, 7], 'c': [8, 9, 10]}\n",
      "2\n",
      "5\n",
      "9\n"
     ]
    }
   ],
   "source": [
    "b = {'a' : [1, 2, 3], 'b' : [4, 5, 7], 'c' : [8, 9, 10]}\n",
    "#test = b.pop(\"a\")\n",
    "#print(test)\n",
    "print(b)\n",
    "\n",
    "for value in b.values():\n",
    "    print(value[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'two_one_three'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def test(num1, num2, num3):\n",
    "    return \"_\".join([num1, num2, num3])\n",
    "\n",
    "test(\"one\", \"two\", \"three\")\n",
    "test(num2 = \"one\", num1 = \"two\", num3 = \"three\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1', '2']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[\"1\", \"2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "forward\n"
     ]
    }
   ],
   "source": [
    "def paired_read_checker(flag):\n",
    "    '''Takes the bitwise flag.\n",
    "    Checks if reads are paired-end and halts program if reads are not.\n",
    "    Checks if reads are mapped in proper pairs and halts program if not.\n",
    "    Checks if the read is forward or reverse and returns \"forward\" or \"reverse\".'''\n",
    "    \n",
    "    # Check if reads are paired\n",
    "    if ((flag & 1) != 1):\n",
    "        raise NameError(\"ERROR: Exiting program. This is not paired-end data.\")\n",
    "    \n",
    "    # Check if reads are mapped in proper pairs\n",
    "    if ((flag & 2) != 2):\n",
    "        raise NameError(\"ERROR: Exiting program. Encountered reads that were not mapped in proper pairs.\")\n",
    "    \n",
    "    # Check forward or reverse\n",
    "    # Flags drived from Sequence Alignment/Map Format Specification (21 Aug 2017)\n",
    "    # https://samtools.github.io/hts-specs/SAMv1.pdf\n",
    "    \n",
    "    if ((flag & 64) == 64):\n",
    "    \tread = \"forward\"\n",
    "    if ((flag & 128) == 128):\n",
    "        read = \"reverse\"\n",
    "\treturn read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created UMI dictionary.\n",
      "Opening SAM files for reading and writing.\n",
      "\n",
      "\n",
      "UMI\tCount\n",
      "AAAAAAAA\t8\n",
      "TTTTTTTT\t5\n",
      "\n",
      "\n",
      "Finished.\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "\n",
    "\n",
    "#####################################################################################\n",
    "##### Define Higher Order Functions #################################################\n",
    "#####################################################################################\n",
    "\n",
    "def strand_checker(flag):\n",
    "    '''Takes the bitwise flag and checks it for strandedness. Assumes read is mapped, otherwise returnes None.\n",
    "    Assumes data are single-end. Returns \"+\" or \"-\", depending on strand.'''\n",
    "    # Check read is mapped\n",
    "    if ((flag & 4) == 4):\n",
    "        return None\n",
    "    # Assume positive strand\n",
    "    strand = \"+\"\n",
    "    if ((flag & 16) == 16):\n",
    "        # Changes strand if bit 16 is set\n",
    "        strand = \"-\"\n",
    "    return strand\n",
    "\n",
    "def POS_correct(cigar_string, POS):\n",
    "    '''Takes the CIGAR string and corrects the start position in the SAM file (POS)\n",
    "    according to any soft clipping at the beginning of the CIGAR string. Returns the corrected POS value.'''\n",
    "    search = re.search(r\"^\\d+S\", cigar_string)\n",
    "    if search:\n",
    "        soft_clip = int(search.group(0)[:-1])\n",
    "        corrected_POS = POS - soft_clip\n",
    "    else:\n",
    "        corrected_POS = POS\n",
    "        # Corrected starting position could be the same as the starting position\n",
    "    return corrected_POS\n",
    "\n",
    "def perbase_qscore(quality):\n",
    "    '''Calculates the average per-base quality score (assumes Phred33 encoding) \n",
    "    based on the QUAL string in the SAM entry.'''\n",
    "    score_total = 0\n",
    "    for char in quality:\n",
    "        score_total += ord(char) - 33 # Convert ASCII score to phred score\n",
    "    score = score_total / len(quality)\n",
    "    return score\n",
    "\n",
    "def count_lines(infile):\n",
    "    '''Opens the input file and returns the number of lines in the file.'''\n",
    "    with open(infile) as file:\n",
    "        for i, line in enumerate(file):\n",
    "            pass\n",
    "    return i + 1\n",
    "\n",
    "\n",
    "\n",
    "#####################################################################################\n",
    "##### Set Up ########################################################################\n",
    "#####################################################################################\n",
    "\n",
    "### Initialize empty dictionaries for tuples and dictionaries\n",
    "tuple_dict = {}\n",
    "umi_dict = {}\n",
    "\n",
    "### Make UMI dictionary\n",
    "with open(\"umi_list.txt\") as umi_list:\n",
    "    for umi in umi_list:\n",
    "        umi = umi.strip(\"\\n\")\n",
    "        umi_dict[umi] = 0\n",
    "print(\"Created UMI dictionary.\", flush = True)\n",
    "\n",
    "###Initialze chromosome variable for chromosome check in the with loop\n",
    "chromosome = 1\n",
    "\n",
    "### Create output filename\n",
    "infile = \"/Users/sally_claridge/Desktop/deduper-sclaridg/test_input.sam\"\n",
    "filename = infile.split(\"/\")[-1].split(\".\")[0]   # Isolate file name\n",
    "deduped = \"\".join([filename, \"_deduped\"])      # Add \"deduped\"\n",
    "outfile = infile.replace(filename, deduped)       # Replace in original filepath\n",
    "\n",
    "### Get file linecount\n",
    "total_lines = count_lines(infile)\n",
    "\n",
    "\n",
    "\n",
    "#####################################################################################\n",
    "##### Read Through SAM File #########################################################\n",
    "#####################################################################################\n",
    "\n",
    "print(\"Opening SAM files for reading and writing.\", flush = True)\n",
    "\n",
    "with open(infile, \"r+\") as file, open(outfile, \"w+\") as out:\n",
    "    linecount = 0\n",
    "    for line in file:\n",
    "        linecount += 1\n",
    "        if line.startswith(\"@\") == True:\n",
    "            pass\n",
    "        elif line.startswith(\"@\") == False:\n",
    "            line_list = line.strip(\"\\n\").split(\"\\t\")\n",
    "            \n",
    "            ### Progress report\n",
    "            if linecount % 200000 == 0:\n",
    "                print(\"Passed line\" + str(linecount) + \".\", flush = True)\n",
    "            \n",
    "            ### Check UMI\n",
    "            umi = line_list[0].split(\":\")[-1]\n",
    "            if umi in umi_dict:\n",
    "                umi_dict[umi] +=1\n",
    "            else:\n",
    "                if linecount == total_lines:\n",
    "                    for value in tuple_dict.values():\n",
    "                        out.write(value)\n",
    "                continue\n",
    "                \n",
    "            ### Check bitwise flag\n",
    "            flag = int(line_list[1])\n",
    "            strand = strand_checker(flag)\n",
    "            \n",
    "            ### Check chromosome\n",
    "            prev_chromosome = chromosome\n",
    "            chromosome = line_list[2]\n",
    "            if chromosome != prev_chromosome:\n",
    "                for value in tuple_dict.values():\n",
    "                    out.write(value)\n",
    "                tuple_dict = {}\n",
    "            \n",
    "            ### Check POS and CIGAR string\n",
    "            POS = int(line_list[3])\n",
    "            cigar_string = line_list[5]\n",
    "            new_POS = POS_correct(cigar_string, POS)\n",
    "            \n",
    "            ### Calculate average quality score across the alignment\n",
    "            quality = line_list[10]\n",
    "            score = perbase_qscore(quality)\n",
    "            \n",
    "            ### Check tuple_dict\n",
    "            query_tuple = (umi, strand, chromosome, new_POS)\n",
    "            if query_tuple in tuple_dict:\n",
    "                existing_quality = tuple_dict[query_tuple].split(\"\\t\")[10]\n",
    "                existing_score = perbase_qscore(existing_quality)\n",
    "                if existing_score >= score:\n",
    "                    pass\n",
    "                elif existing_score < score:\n",
    "                    tuple_dict[query_tuple] = line\n",
    "            if query_tuple not in tuple_dict:\n",
    "                tuple_dict[query_tuple] = line\n",
    "                \n",
    "            ### Last line check\n",
    "            if linecount == total_lines:\n",
    "                for value in tuple_dict.values():\n",
    "                    out.write(value)\n",
    "\n",
    "### Print closing remarks\n",
    "print(\"\\n\")\n",
    "print(\"UMI\\tCount\", flush = True)\n",
    "for k, v in umi_dict.items():\n",
    "    print(str(k) + \"\\t\" + str(v), flush = True)\n",
    "\n",
    "print(\"\\n\") \n",
    "print(\"Finished.\", flush = True)\n",
    "#sys.exit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Syntax with `argparse`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "#claridge_deduper.py\n",
    "\n",
    "import argparse\n",
    "import re\n",
    "import sys\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"Removes PCR duplicates from a SAM file of sorted (using samtools) and uniquely mapped reads (single- and paired-end). For each set of duplicates, the duplicate with the highest per-base average quality is printed to an output file. Given a list of known UMIs, alignments with unexpected UMIs are ignored.\")\n",
    "parser.add_argument('-f','--file', help='absolute path to SAM file to be deduped.', required=True, type=str)\n",
    "parser.add_argument('-p','--paired', help='reads are paired-end.', required=False, action='store_true', default=False)\n",
    "parser.add_argument('-u','--umi', help='absolute path to list of UMIs (default: randomers were used).', required=False, type=str)\n",
    "args = parser.parse_args()\n",
    "\n",
    "\n",
    "#####################################################################################\n",
    "##### Define Higher Order Functions #################################################\n",
    "#####################################################################################\n",
    "\n",
    "def strand_checker(flag):\n",
    "    '''Takes the bitwise flag and checks it for strandedness. Assumes read is mapped, otherwise returnes None.\n",
    "    Assumes data are single-end. Returns \"+\" or \"-\", depending on strand.'''\n",
    "    # Check read is mapped\n",
    "    if ((flag & 4) == 4):\n",
    "        raise NameError(\"Exiting program. Read is unmapped.\")\n",
    "        return None\n",
    "    # Assume positive strand\n",
    "    strand = \"+\"\n",
    "    if ((flag & 16) == 16):\n",
    "        # Changes strand if bit 16 is set\n",
    "        strand = \"-\"\n",
    "    return strand\n",
    "\n",
    "def read_checker(flag):\n",
    "    '''Takes the bitwise flag and checks whether the reads are paired-end or not.\n",
    "    Halts program if reads are unpaired. Returns \"forward\" or \"reverse\", depending on the read.'''\n",
    "    # Check if reads are paired\n",
    "    if ((flag & 1) != 1):\n",
    "        raise NameError(\"Exiting program. This is not paried-end data.\")\n",
    "        return None\n",
    "    if ((flag & 40) == 40):\n",
    "        read = \"forward\"\n",
    "    if ((flag & 80) == 80):\n",
    "        read = \"reverse\"\n",
    "    return read\n",
    "\n",
    "def POS_correct(cigar_string, POS):\n",
    "    '''Takes the CIGAR string and corrects the start position in the SAM file (POS)\n",
    "    according to any soft clipping at the beginning of the CIGAR string. Returns the corrected POS value.'''\n",
    "    # Find soft clippling in begining of cigar string\n",
    "    search = re.search(r\"^\\d+S\", cigar_string)\n",
    "    # If soft clipping found\n",
    "    if search:\n",
    "        soft_clip = int(search.group(0)[:-1])\n",
    "        corrected_POS = POS - soft_clip\n",
    "    # If soft clipping not found\n",
    "    else:\n",
    "        corrected_POS = POS\n",
    "        # Corrected starting position could be the same as the starting position\n",
    "    return corrected_POS\n",
    "\n",
    "def perbase_qscore(quality):\n",
    "    '''Calculates the average per-base quality score (assumes Phred33 encoding) \n",
    "    based on the QUAL string in the SAM entry.'''\n",
    "    score_total = 0\n",
    "    for char in quality:\n",
    "        score_total += ord(char) - 33 # Convert ASCII score to phred score\n",
    "    score = score_total / len(quality)\n",
    "    return score\n",
    "\n",
    "def count_lines(infile):\n",
    "    '''Opens the input file and returns the number of lines in the file.'''\n",
    "    with open(infile) as file:\n",
    "        for i, line in enumerate(file):\n",
    "            pass\n",
    "    return i + 1\n",
    "\n",
    "\n",
    "\n",
    "#####################################################################################\n",
    "##### Set Up ########################################################################\n",
    "#####################################################################################\n",
    "\n",
    "### Initialize empty dictionaries for tuples and dictionaries\n",
    "# tuple_dict will contain tuples as keys and SAM entries with highest average per-base quality as values\n",
    "# Format of tuples: (umi/randomer, strand +/-, chromosome, new_POS, read forward/reverse/na)\n",
    "tuple_dict = {}\n",
    "# umi_dict will store UMIs or randomers as keys and counts as values\n",
    "umi_dict = {}\n",
    "# Make UMI dictionary if UMI flag is set, otherwise leave empty\n",
    "if args.umi is not None:\n",
    "    with open(args.umi) as umi_list:\n",
    "        for umi in umi_list:\n",
    "            umi = umi.strip(\"\\n\")\n",
    "            umi_dict[umi] = 0\n",
    "    print(\"Created UMI dictionary.\", flush = True)\n",
    "\n",
    "### Initialze first_line and read variables\n",
    "# Used in chromosome check\n",
    "first_line = True\n",
    "# Read will be a non-factor in single-end data, but will be edited if paired flag set\n",
    "read = \"na\"\n",
    "\n",
    "### Create output filename\n",
    "infile = \"/Users/sally_claridge/Desktop/deduper-sclaridg/test_input.sam\"\n",
    "filename = infile.split(\"/\")[-1].split(\".\")[0]   # Isolate file name\n",
    "deduped = \"\".join([filename, \"_deduped\"])        # Add \"deduped\"\n",
    "outfile = infile.replace(filename, deduped)      # Replace in original filepath\n",
    "\n",
    "### Get file linecount to be used in last line check\n",
    "total_lines = count_lines(infile)\n",
    "\n",
    "\n",
    "\n",
    "#####################################################################################\n",
    "##### Read Through SAM File #########################################################\n",
    "#####################################################################################\n",
    "\n",
    "print(\"Opening SAM files for reading and writing.\", flush = True)\n",
    "\n",
    "with open(infile, \"r+\") as file, open(outfile, \"w+\") as out:\n",
    "    linecount = 0\n",
    "    for line in file:\n",
    "        linecount += 1\n",
    "        if line.startswith(\"@\") == True:\n",
    "            pass\n",
    "        elif line.startswith(\"@\") == False:\n",
    "            line_list = line.strip(\"\\n\").split(\"\\t\")\n",
    "            \n",
    "            ### Progress report\n",
    "            if linecount % 200000 == 0:\n",
    "                print(\"Passed line\" + str(linecount) + \".\", flush = True)\n",
    "            \n",
    "            ### Check UMI\n",
    "            umi = line_list[0].split(\":\")[-1]\n",
    "            if args.umi is None:                  # If UMI flag was not set, assume randomers\n",
    "                if umi in umi_dict:\n",
    "                    umi_dict[umi] +=1\n",
    "                else:\n",
    "                    umi_dict[umi] = 1             # Still use umi_dict, but store randomers instead of UMIs\n",
    "            elif args.umi is not None:            # If UMI flag was set\n",
    "                if umi in umi_dict:\n",
    "                    umi_dict[umi] +=1\n",
    "            else:\n",
    "                if linecount == total_lines:      # If last line in file, print dictionary contents\n",
    "                    for value in tuple_dict.values():\n",
    "                        out.write(value)\n",
    "                continue                          # Do not write unexpexted UMI alignment to file\n",
    "                \n",
    "            ### Check bitwise flag\n",
    "            flag = int(line_list[1])\n",
    "            # Check strand\n",
    "            strand = strand_checker(flag)\n",
    "            # Check read (if paired flag set)\n",
    "            if args.paired:\n",
    "                read = read_checker(flag)\n",
    "                \n",
    "            ### Check chromosome\n",
    "            # If first line, set the chromosome variable\n",
    "            if first_line == True:\n",
    "                chromosome = line_list[2]\n",
    "                first_line == False\n",
    "            # If not the first line, run chromosome check for emptying tuple_dict\n",
    "            elif first_line == False:\n",
    "                prev_chromosome = chromosome\n",
    "                chromosome = line_list[2]\n",
    "                if chromosome != prev_chromosome:  # Inequality means that you have switched chromosomes\n",
    "                    for value in tuple_dict.values():\n",
    "                        out.write(value)           # Print tuple_dict contents to output file\n",
    "                    tuple_dict = {}                # Empty tuple_dict\n",
    "            \n",
    "            ### Check POS and CIGAR string (correcting for soft clipping if necessary)\n",
    "            POS = int(line_list[3])\n",
    "            cigar_string = line_list[5]\n",
    "            new_POS = POS_correct(cigar_string, POS)\n",
    "            \n",
    "            ### Calculate average per-base quality score across the alignment\n",
    "            quality = line_list[10]\n",
    "            score = perbase_qscore(quality)\n",
    "            \n",
    "            ### Check tuple_dict\n",
    "            # Make tuple\n",
    "            query_tuple = (umi, strand, chromosome, new_POS, read)\n",
    "            if query_tuple in tuple_dict:\n",
    "                # Check existing score against new alignment's score, replace if new alignments's is higher\n",
    "                existing_quality = tuple_dict[query_tuple].split(\"\\t\")[10]\n",
    "                existing_score = perbase_qscore(existing_quality)\n",
    "                if existing_score >= score:\n",
    "                    pass\n",
    "                elif existing_score < score:\n",
    "                    tuple_dict[query_tuple] = line\n",
    "            if query_tuple not in tuple_dict:\n",
    "                # Add alignment if tuple not found in tuple_dict\n",
    "                tuple_dict[query_tuple] = line\n",
    "                \n",
    "            ### Last line check\n",
    "            if linecount == total_lines:\n",
    "                for value in tuple_dict.values():\n",
    "                    out.write(value)\n",
    "\n",
    "### Print closing remarks/stats       \n",
    "print(\"\\n\")\n",
    "if args.umi is None:\n",
    "    print(\"%s randomers were identified.\" % len(umi_dict), flush = True)\n",
    "elif args.umi is not None:\n",
    "    print(\"Molecular_Identifier\\tCount\", flush = True)\n",
    "    for k, v in umi_dict.items():\n",
    "        print(str(k) + \"\\t\" + str(v), flush = True)\n",
    "\n",
    "print(\"\\n\") \n",
    "print(\"Finished.\", flush = True)\n",
    "sys.exit()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
