---
title: "De-Duper"
author: "Sally Claridge"
date: "18 October 2017"
output: html_document
---


Part 1
================================================================================

### Define the Problem

PCR duplicates are two or more reads that were read from the same molecule in the input library. Duplicates should theoretically have the exact same sequence and alignment position as each other. If there are many PCR duplicates present in a dataset and they aren't filtered out, they could bias analyses of relative transcript abundance by artificially inflating the number of reads that came from one particular transcript. PCR duplicates could also overwhelm the signal of rare transcripts, or if one allele amplifies more quickly than the other, the duplicates affect your interpretations of variant calling or heterozygosity.

### Strategy Overview

- This algorithm will take in a SAM file of uniquely mapped, single-end reads and remove PCR duplicates.
- **Sort**: Start by sorting the SAM file with `samtools sort`, which sorts the alignments by chromosome and then leftmost position.
- **Check UMI**: To filter outs reads with UMIs containing errors, extract the UMI from read name and compare it to the whitelist of 96 expected UMIs. If the UMI is not in the list, pass the read and don't do anything. If the UMI exists, save the UMI and look at the chromosome.
- **Check Chromosome**: Duplicates should be on the same chromosome. Save the chromosome, then look at the strand.
- **Check Strand**: Duplicates should be on the same strand. Save the strand, then look at starting position.
- **Correct Start Position**: If the cigar string contains an S, minus the number preceding the S (e.g. `^[0-9]+S`) from the start position of that alignment. Save this new start position. If the cigar string doesn't contain an S, save the starting position as-is.
- **Increment Dictionary**: Make a tuple of the UMI, strand, chromosome, and start position (e.g. `(AAAAAAAA, 0, 14, 100000000)`). If the tuple exists in your dictionary, increment the counter and pass. This means that this is probably a PCR duplicate. If the tuple doesn't exist, intialize it at a count of 1 and write the associated alignment to the output file.

### Example SAM File

- There are two expected UMIs in this sample file: `AAAAAAAA` and `TTTTTTTT`
- The file contains 15 SAM records (no SAM header)
- Assume that thhis file has already been put through `samtools sort`

```
query_information:AAAAAAAA	0	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	0	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAATAA	0	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:TTTTTTTT	0	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	16	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	16	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:TTTTTTTT	0	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:TTTTTTTT	0	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	0	1	67240002	36	2S68M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:TTTTTTTT	0	1	67240004	36	4S66M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	0	1	67240010	36	5S65M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	0	1	67275000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:TTTTTTTT	16	1	76540000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	0	2	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:TTTTTATT	16	2	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
```

Below is a breakdown of the 15 records and their expected tuples. There should be 7 records in the output file, denoted by "First instance of this tuple".

1. `(AAAAAAAA, 0, 1, 67240000)`: First instance of this tuple
2. `(AAAAAAAA, 0, 1, 67240000)`: Duplicate of Record 1
3. `(AAAAATAA, 0, 1, 67240000)`: UMI contains error
4. `(TTTTTTTT, 0, 1, 67240000)`: First instance of this tuple
5. `(AAAAAAAA, 16, 1, 67240000)`: First instance of this tuple
6. `(AAAAAAAA, 16, 1, 67240000)`: Duplicate of Record 6
7. `(TTTTTTTT, 0, 1, 67240000)`: Duplicate of Record 4
8. `(TTTTTTTT, 0, 1, 67240000)`: Duplicate of Record 4
9. `(AAAAAAAA, 0, 1, 67240000)`: Duplicate of Record 1 (soft-clipped by 2)
10. `(TTTTTTTT, 0, 1, 67240000)`: Duplicate of Record 4(soft-clipped by 4)
11. `(AAAAAAAA, 0, 1, 67240005)`: First instance of this tuple (soft-clipped by 5)
12. `(AAAAAAAA, 0, 1, 67275000)`: First instance of this tuple
13. `(TTTTTTTT, 16, 1, 76540000)`: First instance of this tuple
14. `(AAAAAAAA, 0, 2, 67240000)`: First instance of this tuple
15. `(TTTTTATT, 16, 2, 67240000)`: UMI contains error


### Pseudocode

First, `samtools sort` the file on the command line.

```
open umi list
    make dictionary of known UMIs (e.g. umi_dict), counts of 0 as the values

make empty dictionary for tuples (e.g. tuple_dict)

initialize chromosome variable (see "prev_chrom" check below)

open file:
    for each alignment:
        # check UMI
        if UMI in umi_dict:
            increment counter
            save UMI
        else:
            pass and go to next alignment (break)
        save strand (from bitwise flag, 0 or 16)
        isolate chromosome, for chromosome:
            save previous chromosome to new variable prev_chrom
            save new chromosome
            if this new chromosome is not the same as the prev_chrom:
                empty the tuple_dict
            if this chromosome is the same as the prev_chrom:
                pass
        save starting position
        save cigar string
        if cigar string doesn't contain soft clipping notation:
            make tuple of UMI, chromosome, strand, starting position
            if tuple is in tuple_dict:
                increment and pass
            else if tuple is not in tuple_dict:
                initialize tuple key with value as 1
                write current alignment to output file
        else if cigar_string contains soft clipping notation:
            save number preceding "S" (clipped_num)
            minus clipped_num from starting position (new_start_pos)
            make tuple of UMI, chromosom, strand, new_start_pos
            if tuple is in tuple_dict:
                increment and pass
            else if tuple is not in tuple_dict:
                initialize tuple key with value as 1
                write current alignment to output file
```


### Expected Output

There are 7 alignments remaining:

```
query_information:AAAAAAAA	0	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:TTTTTTTT	0	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	16	1	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	0	1	67240010	36	5S65M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	0	1	67275000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:TTTTTTTT	16	1	76540000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
query_information:AAAAAAAA	0	2	67240000	36	70M	*	0	0	AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA	GCEEAEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
```







